src-js/fontra-webcomponents/src/custom-data-list.js

```diff
@@ -0,0 +1,347 @@
+import { zip } from "@fontra/core/utils.js";
+import { dialogSetup, message } from "@fontra/web-components/modal-dialog.js";
+import { UIList } from "@fontra/web-components/ui-list.js";
+
+// TODO: Refactor this, copy from panel-axes.js
+function arraysEqual(arrayA, arrayB) {
+  if (arrayA.length !== arrayB.length) {
+    return false;
+  }
+  for (const [itemA, itemB] of zip(arrayA, arrayB)) {
+    if (itemA !== itemB) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// TODO: Refactor this, copy from panel-axes.js
+function updateRemoveButton(list, buttons) {
+  list.addEventListener("listSelectionChanged", (event) => {
+    buttons.disableRemoveButton = list.getSelectedItemIndex() === undefined;
+  });
+}
+
+export class CustomDataList extends SimpleElement {
+  static styles = `
+
+  .fontra-ui-font-info-sources-panel-list-element {
+    min-width: max-content;
+    max-width: 29.5em; // 4.5 + 25
+    max-height: 12em;
+  }
+  `;
+
+  constructor(controller, customDataInfos) {
+    super();
+    this.contentElement = this.shadowRoot.appendChild(html.div());
+    this.controller = controller;
+    this.customDataInfos = customDataInfos;
+    this.render();
+  }
+
+  buildCustomDataList() {
+    const customDataNames = this.customDataInfos.map((info) => info.key);
+    const model = this.controller.model;
+
+    const makeItem = ([key, value]) => {
+      const item = new ObservableController({ key: key, value: value });
+      item.addListener((event) => {
+        const sortedItems = [...labelList.items];
+        sortedItems.sort(
+          (a, b) =>
+            (customDataNames.indexOf(a.key) != -1
+              ? customDataNames.indexOf(a.key)
+              : customDataNames.length) -
+            (customDataNames.indexOf(b.key) != -1
+              ? customDataNames.indexOf(b.key)
+              : customDataNames.length)
+        );
+
+        if (!arraysEqual(labelList.items, sortedItems)) {
+          labelList.setItems(sortedItems);
+        }
+
+        const newCustomData = sortedItems.map((customData) => {
+          return { ...customData };
+        });
+        model.customData = newCustomData;
+      });
+      return item.model;
+    };
+
+    const sortedItems = Object.entries(model);
+    sortedItems.sort(
+      (a, b) =>
+        (customDataNames.indexOf(a[0]) != -1
+          ? customDataNames.indexOf(a[0])
+          : customDataNames.length) -
+        (customDataNames.indexOf(b[0]) != -1
+          ? customDataNames.indexOf(b[0])
+          : customDataNames.length)
+    );
+    const items = sortedItems?.map(makeItem) || [];
+
+    const labelList = new UIList();
+    labelList.classList.add("fontra-ui-font-info-sources-panel-list-element");
+    labelList.style = `min-width: 12em;`;
+    labelList.columnDescriptions = [
+      {
+        key: "key",
+        title: "Key", // TODO: translation
+        width: "14em",
+        editable: true,
```

Comment:
I don't think `editing` is useful, given we restrict ourselves to known keys.

Reply:
Also, if I dismiss the dialog, the key goes into edit mode, which isn't desirable, either.

Reply:
Thanks. Done.